cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaLogWriterPlugin.java

```diff
@@ -0,0 +1,247 @@
+      long key = logEvent.getTimeStamp() / eventBucketIntervalMs;
+
+      // Sleep while we can add the entry
+      while (true) {
+        // Get the oldest bucket in the table
+        long oldestBucketKey = 0;
+        synchronized (messageTable) {
+          SortedSet<Long> rowKeySet = messageTable.rowKeySet();
+          if (rowKeySet.isEmpty()) {
+            // Table is empty so go ahead and add the current event in the table
+            break;
+          }
+          oldestBucketKey = rowKeySet.first();
+        }
+
+        // If the current event falls in the bucket number which is not in window [oldestBucketKey, oldestBucketKey+8]
+        // sleep for the time duration till event falls in the window
+        if (key > (oldestBucketKey + maxNumberOfBucketsInTable)) {
+          LOG.trace("key={}, oldestBucketKey={}, maxNumberOfBucketsInTable={}. Sleeping for {} ms.",
+                    key, oldestBucketKey, maxNumberOfBucketsInTable, SLEEP_TIME_MS);
+
+          if (countDownLatch.await(SLEEP_TIME_MS, TimeUnit.MILLISECONDS)) {
+            // if count down occurred return
+            LOG.info("Returning since callback is cancelled");
+            return;
+          }
+
+        } else {
+          break;
+        }
+      }
+
+      synchronized (messageTable) {
+        Map.Entry<Long, List<KafkaLogEvent>> entry = messageTable.get(key,
+                                                                      loggingContext.getLogPathFragment(logBaseDir));
+        List<KafkaLogEvent> msgList = null;
+        if (entry == null) {
+          long eventArrivalBucketKey = System.currentTimeMillis() / eventBucketIntervalMs;
+          msgList = Lists.newArrayList();
+          messageTable.put(key, loggingContext.getLogPathFragment(logBaseDir),
+                           new AbstractMap.SimpleEntry<Long, List<KafkaLogEvent>>(eventArrivalBucketKey, msgList));
+        } else {
+          msgList = messageTable.get(key, loggingContext.getLogPathFragment(logBaseDir)).getValue();
+        }
+        msgList.add(new KafkaLogEvent(event.getGenericRecord(), event.getLogEvent(), loggingContext,
+                                      event.getPartition(), event.getNextOffset()));
+      }
+    } catch (Throwable th) {
+      LOG.warn("Exception while processing message with nextOffset {}. Skipping it.", event.getNextOffset(), th);
+    }
+  }
+
+  @Override
+  public void stop() {
+    try {
+      logFileWriter.flush();
+      logFileWriter.close();
+
+      if (logWriterFuture != null && !logWriterFuture.isCancelled() && !logWriterFuture.isDone()) {
+        logWriterFuture.cancel(false);
+        logWriterFuture = null;
+      }
+
+      if (cleanupFuture != null && !cleanupFuture.isCancelled() && !cleanupFuture.isDone()) {
+        cleanupFuture.cancel(false);
+        cleanupFuture = null;
+      }
+
+      countDownLatch.countDown();
+
+      if (scheduledExecutor != null) {
+        scheduledExecutor.shutdown();
+      }
```

Comment:
Although this is not part of this PR, I would suggest to rearrange the stop logic to make it more resilient.
```java
if (countDownLatch != null) {
  countDownLatch.countDown();
}

if (scheduledExecutor != null) {
  scheduledExecutor.shutdown();
}
scheduledExecutor.awaitTermination(5, TimeUnit.MINUTES);

logFileWriter.flush();
logFileWriter.close();
```

We don't need `logWriterFuture` and `cleanupFuture` in this case.